### PR TITLE
8256390: ZGC: Relocate in-place instead of having a heap reserve

### PR DESCRIPTION
**Pull request by pliden**
The heap reserve is a portion of the heap that is set aside to cope with “emergency situations”, like when ZGC needs to compact the heap when it’s already full. However, due to a number of factors (mutator relocation, page pinning, etc) there is no guarantee that relocation will always succeed. In fact, we can end up in situations where we fail to reclaim memory even when 90% of the heap is garbage (JDK-8254346 is an example of where this happens).

So, even if the heap reserve serves its purpose well in most cases, it still doesn't offer any guarantees that relocation succeeds in all cases. To get the guarantees we want we could, instead of having a heap reserve, turn to doing in-place relocation. In other words, if a worker fails to allocate a new page, then it compacts the page currently being relocated, and then use that page as the target for new allocations. In addition, if a mutator fails to relocate an object, then it waits for a worker to relocate the object, instead of pinning the page. This patch does exactly that, and hence makes ZGC more resilient in situations where memory is scares. It also means that ZGC works even better when using small heap (where memory tends to be scares).

Testing: Tier1-7 on linux-x64, Tier-1-3 on all Oracle platforms, 50+ runs of gc-test-suite on linux-x64 (both with and without `-XX:+ZStressRelocateInPlace`).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ❌ (1/6 failed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test tasks**
- [Linux x64 (build hotspot zero)](https://github.com/pliden/jdk/runs/1444162221)
- [Linux x86 (build debug)](https://github.com/pliden/jdk/runs/1444162135)
- [Linux x86 (build release)](https://github.com/pliden/jdk/runs/1444161472)

### Issue
 * [JDK-8256390](https://bugs.openjdk.java.net/browse/JDK-8256390): ZGC: Relocate in-place instead of having a heap reserve

### Reviewers
 * [Stefan Karlsson](https://openjdk.java.net/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to 106b4dfd1f7aaf5181a03430486a1abaf01a0452
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 106b4dfd1f7aaf5181a03430486a1abaf01a0452

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1228/head:pull/1228`
`$ git checkout pull/1228`

### CODE REVIEW

File: src/hotspot/share/gc/z/zPageAllocator.cpp [hunk 1]
```diff
@@ -259,42 +255,30 @@ size_t ZPageAllocator::capacity() const {
...
 
 size_t ZPageAllocator::unused() const {
-  const ssize_t capacity = (ssize_t)Atomic::load(&_capacity);
-  const ssize_t used = (ssize_t)Atomic::load(&_used);
-  const ssize_t claimed = (ssize_t)Atomic::load(&_claimed);
-  const ssize_t max_reserve = (ssize_t)_max_reserve;
-  const ssize_t unused = capacity - used - claimed - max_reserve;
-  return unused > 0 ? (size_t)unused : 0;
+  const ZPageAllocatorStats st = stats();
+  return st.capacity() - st.used() - st.claimed();
```
Comment (stefank) [inline src/hotspot/share/gc/z/zPageAllocator.cpp:264]: Used by JVM_FreeMemory. I'd prefer if that usage was kept lock free.

File: src/hotspot/share/gc/z/zPageAllocator.cpp [hunk 2]
```diff
@@ -793,7 +747,7 @@ size_t ZPageAllocator::uncommit(uint64_t* timeout) {
     }
 
     // Record flushed pages as claimed
-    Atomic::add(&_claimed, flushed);
+    _claimed += flushed;
```
Comment (stefank) [inline src/hotspot/share/gc/z/zPageAllocator.cpp:750]: See FreeMemory comment above.

**Review by stefank — CHANGES_REQUESTED** [review]
First pass looking through the patch

**Comment by pliden** [comment]
@stefank Thanks for reviewing! All of your comments above have been addressed.

**Review by fisk — CHANGES_REQUESTED** [review]
Looks awesome in general. Getting rid of the reserve and making sure relocation can't fail is indeed very nice. One question mark only about the _ref_count and lack of load_acquire in double-checked locking patterns. I think we should load_acquire also inside of the lock, so we don't rely on the release of the lock also being an acquire. Unless I missed something in the protocol.

File: src/hotspot/share/gc/z/zStat.hpp [hunk 3]
```diff
@@ -483,82 +487,61 @@ class ZStatHeap : public AllStatic {
...
+    size_t reclaimed;
   } _at_relocate_end;
 
   static size_t capacity_high();
   static size_t capacity_low();
-  static size_t available(size_t used);
-  static size_t reserve(size_t used);
   static size_t free(size_t used);
+  static size_t allocated(size_t used, size_t reclaiemd);
+  static size_t garbage(size_t reclaiemd);
```
Comment (stefank) [inline src/hotspot/share/gc/z/zStat.hpp:536]: reclaiemd -> reclaimed

File: src/hotspot/share/gc/z/zStat.cpp [hunk 4]
```diff
@@ -1270,88 +1273,65 @@ size_t ZStatHeap::capacity_low() {
...
 
-size_t ZStatHeap::available(size_t used) {
+size_t ZStatHeap::free(size_t used) {
   return _at_initialize.max_capacity - used;
 }
 
-size_t ZStatHeap::reserve(size_t used) {
-  return MIN2(_at_initialize.max_reserve, available(used));
+size_t ZStatHeap::allocated(size_t used, size_t reclaimed) {
+  return (used + reclaimed) - _at_mark_start.used;
```
Comment (stefank) [inline src/hotspot/share/gc/z/zStat.cpp:1281]: I don't know why used + reclaimed is the amount allocated. Is this correct? Could you add a comment?

File: src/hotspot/share/gc/z/zForwarding.cpp [hunk 5]
```diff
@@ -22,14 +22,144 @@
...
+
+    return;
+  }
+}
+
+void ZForwarding::wait_page_released() const {
+  if (Atomic::load_acquire(&_ref_count) != 0) {
+    ZStatTimer timer(ZCriticalPhaseRelocationStall);
+    ZLocker<ZConditionLock> locker(&_ref_lock);
+    while (Atomic::load(&_ref_count) != 0) {
```
[the remark this answers is not in the document]
Comment (pliden) [reply src/hotspot/share/gc/z/zForwarding.cpp:136]: Will fix!

File: src/hotspot/share/gc/z/zStat.hpp (hunk 3 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zStat.hpp:536]: Will fix!

File: src/hotspot/share/gc/z/zStat.cpp (hunk 4 above)
Comment (pliden) [reply src/hotspot/share/gc/z/zStat.cpp:1281]: We do "used_now - used_at_some_earlier_point" to see how much the used grew (i.e. now much was allocated). However, if we just do that, we will fail to take into account that we might also have reclaimed memory, which would have lowered "used_now". So, we let "used_now" be "used + reclaimed" to get the used we would have seen if we had not reclaimed any memory. Hope that makes sense. I'll add a comment.

**Comment by pliden** [comment]
@stefank @fisk Pushed an update, which addresses your comments above. Also merged with master to resolve a conflict.

**Review by stefank — APPROVED** [review]

**Comment by albertnetymk** [comment]
Not a review, just a comment in passing

I believe by merging `relocate_object` and `relocate_object_inner`, one can defer `retain_page()` after the forwarding table returns nothing. This may reduce the blocking window in `detach_page` before freeing a page, since some pages may never be retained.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by pliden** [comment]
@albertnetymk Yes, that would be possible. Let's do that as a follow up, if we think it's worth it.

**Comment by albertnetymk** [comment]
> Yes, that would be possible. Let's do that as a follow up, if we think it's worth it.

Sure; I actually think it would make the code easier to read.

**Comment by pliden** [comment]
Thanks for reviewing!

/integrate

**Pull request closed without merging**